src/charm.py

```diff
@@ -24,36 +29,73 @@ def __init__(self, *args: typing.Any):
             args: Arguments passed to the CharmBase parent constructor.
         """
         super().__init__(*args)
-        self.irc = IRCService()
+        self._irc = IRCBridgeService()
+        self._database = DatabaseObserver(self, DATABASE_RELATION_NAME)
+        self._matrix = MatrixObserver(self, MATRIX_RELATION_NAME)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
         self.framework.observe(self.on.install, self._on_install)
         self.framework.observe(self.on.start, self._on_start)
         self.framework.observe(self.on.stop, self._on_stop)
-        self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
```

Comment:
Shouldn't we keep listening to that event to ensure the snap is upgraded ?

Reply:
Good point, added!